File: routes/stats.js
import express from 'express'
import cors from 'cors'
import { k8sClients } from '../common.js'

const routes = express.Router()

async function getTotalPodCount() {
    const podsPromises = k8sClients.map(client => client.api.v1.pods.get())

    const pods = await Promise.all(podsPromises)
    const podCount = pods.map(podList => podList.body.items.length)

    return podCount.reduce((acc, podCount) => acc + podCount, 0)
}

async function getKubernetesStats() {
    const podCount = await getTotalPodCount()

    return {
        podCount: podCount
    }
}


routes.get('/stats', cors(), async (req, res) => {
    const k8sStatsPromise = getKubernetesStats()

    const [k8sStats] = await Promise.all([k8sStatsPromise])

    res.status(200).json({
        k8s: k8sStats
    })
})

export default routes

File: models/dto/body/stats.go
package body

import "time"

type K8sStats struct {
	PodCount int `json:"podCount" bson:"podCount"`
}

type Stats struct {
	K8sStats K8sStats `json:"k8s" bson:"k8s"`
}

type TimestampedStats struct {
	Stats     Stats     `json:"stats" bson:"stats"`
	Timestamp time.Time `json:"timestamp" bson:"timestamp"`
}
